### ISSUE TITLE
Make CKEditor 5 packages valid ESM

### ISSUE THREAD

**Issue by pomek** [ckeditor/ckeditor5#13673]
## 📝 Provide a description of the improvement

To use CKEditor 5 sources in Node without processing it by any bundler, we can mark our packages as `type:module` in `package.json`.

> There's another issue that is just quasy rellated to this:
> 
> When using ckeditor from sources without a build, vitest (being esm-first project) complains about CK5 modules not being esm modules, although they seem to should be in its opinion. 
> 
> ```
> Module /Users/sitecore/Sites/feaas-components/ckeditor5/node_modules/@ckeditor/ckeditor5-utils/src/dom/findclosestscrollableancestor.js:14 
> seems to be an ES Module but shipped in a CommonJS package. 
> You might want to create an issue to the package "@ckeditor/ckeditor5-utils" 
> asking them to ship the file in .mjs extension or add "type": "module" in their package.json.
> ```
> 
> Is this something that can be resolved?

_Originally posted by @Inviz in https://github.com/ckeditor/ckeditor5/issues/11704#issuecomment-1467413888_

As it is unrelated to TypeScript migration (our sources were modules before migration), I extracted the discussion to another issue.
            
---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

**Comment by Inviz** [comment]
Thanks @pomek . It does seem like that would resolve the problem. It makes sense too.

**Comment by HaNdTriX** [comment]
Is there any workaround regarding this issue?
Adding 

**vite.config.ts**:
```diff
  test: {
    ...
+    deps: {
+      inline: [
+        'ckeditor5'
+      ]
+    }
  }
```

Doesn't seem to work.

**Comment by Witoso** [comment]
Noticed the same, I'm not sure why inlining `ckeditor5` is ignored in for example following config:
```js
export default {
  test: {
    deps: {
      inline: [
        "ckeditor5",
        /ckeditor5/
      ],
    },
  },
};

```

**Comment by filipsobol** [comment]
Adding `"type": "module"` changes not only how our package is consumed, but also how our tools and Node interpret and execute the code.

With `"type": "module"` all `.js` files are expected to use ESM. This causes problems with the webpack config file which uses CJS. Renaming all `webpack.config.js` files to `webpack.config.cjs` fixed this.

Another problem was caused by webpack requiring file extensions in imports in manual and automated tests.

Manual tests can be fixed by updating the `getJavaScriptLoader` function in `@ckeditor/ckeditor5-dev-utils` to return the following object:

```diff
return {}
  test: /\.js$/,
  getDebugLoader( debugFlags ),
+ resolve: {
+   fullySpecified: false,
+ },
};
```

The same problem can be fixed for automated tests by adding the following objects to `module.rules` in `lib/utils/automated-tests/getwebpackconfig.js`:

```js
{
  test: /\.m?js$/,
  type: "javascript/auto",
},
{
  test: /\.m?js$/,
  resolve: {
    fullySpecified: false,
  },
}
```

Unfortunately, this doesn't solve all problems in automated testing - webpack will throw a bunch of errors that it can't find node-specific modules (e.g. `Can't resolve 'stream'`). More research is needed to resolve this issue.

Alternatively, we could rename all tests to `.ts`, as TypeScript files don't seem to suffer from this problem, but this could introduce a bunch of other issues.

For now, the easier way would probably be to add `"type": "module"` to `package.json` during the release process. If we go that route, we need to check if we also need to add `"exports"` to `package.json` to access files other than the main entry points (for example, to access the `src` and `theme` folders or `package.json`).

**Comment by filipsobol** [comment]
The errors with node-specific modules might actually be caused by the setup of my local environment. I'll test this further next week, as the steps above may be enough to get it working.

**Comment by filipsobol** [comment]
I prepared two PRs:

* One to not require file extension in import paths in automated and manual tests (webpack requires them with `type:module`): https://github.com/ckeditor/ckeditor5-dev/pull/885.
* Another to add `type:module` and rename `webpack.config.js` to `webpack.config.cjs`: https://github.com/ckeditor/ckeditor5/pull/14219 (depends on the first one).

TypeScript files don't need file extensions in import paths because we use `"module":"es6"` instead of `"module":"node2016"` in `tsconfig.json`.

I'd consider releasing alpha/beta/nightly/whatever so folks using Vite can test it. This will help us make sure we don't break anything for `39.0.0`. Any thoughts? @Witoso @pomek

**Comment by pomek** [comment]
Changes in cke5-dev need tests not to break things in the future.

**Comment by filipsobol** [comment]
> Changes in cke5-dev need tests not to break things in the future.

Done :heavy_check_mark:

**Comment by pomek** [comment]
Ups, sorry :D

**Comment by filipsobol** [comment]
Four PRs connected to this issue will allow us to release valid ESM packages.

Unfortunately, the changes affect almost our entire codebase, as I  had to:
* add `"type": "module"` to the `package.json` files,
* rename some configuration files from `webpack.config.js` to `webpack.config.cjs`,
* update all imports to include file extensions as required by the ESM spec.

In total, there are about 3000 files changed, which will make these PRs time consuming to review. Unfortunately, the TypeScript team has decided not to introduce any configuration flags that would automatically add missing file extensions to imports/exports, as they want TypeScript to be more of a linter than a compiler in the future.

Additionally, I had to make small adjustments to the webpack configuration (which depends on features introduced about a year ago in `webpack@v5.74.0`) in various places, so we need to make sure that everything still works as expected. So far I've tested:
- [X] Manual tests (`yarn manual`).
- [X] Automated tests (`yarn test`).
- [X] Documentation generation (`yarn docs --production`).
- [X] Part of the release process (`yarn release:prepare-packages --nightly`).
- [X] Using one of the packages in Vitest, which worked correctly (see issues reported to #14097).

An alternative approach would be to use a bundler (like `esbuild`, `rollup`, `vite`, `webpack`, whatever) to build the plugins and continue to use `tsc` just to generate types. This way, we could generate a single `.js' file for each plugin, benefit from tree shaking and code optimizations, and not worry about import/export paths. However, this would be a major breaking change because we would no longer allow importing from `/src`.

Internally there are some concerns due to the size of the changes and the impact they might have and we need to discuss what our next step should be. Personally, I feel that even though the JavaScript ecosystem is a mess, the tools are in a "good enough" state to allow us to fix this. After all, there are more and more ESM-first or even ESM-only packages gaining traction (a good example is Vite and its ecosystem).

**Comment by filipsobol** [comment]
The `v38.2.0-alpha.1` seems to be working fine and be valid ESM package.

Before: [@ckeditor/ckeditor5-utils@38.1.1](https://arethetypeswrong.github.io/?p=%40ckeditor%2Fckeditor5-utils%4038.1.1) - notice the `Unexpected module syntax` error.
After: [@ckeditor/ckeditor5-utils@38.2.0-alpha.1](https://arethetypeswrong.github.io/?p=%40ckeditor%2Fckeditor5-utils%4038.2.0-alpha.1) - the warning indicates that the package is ESM-only and CJS consumers have to use dynamic `import()`. This is expected.

**Comment by filipsobol** [comment]
Changes to make our packages valid ESM are merged and will be released in the January.

Next week I will test our nightly releases in Vite and Vitest.

**Comment by filipsobol** [comment]
The nightly releases seem to be working fine and be a valid ESM package.

### arethetypeswrong.github.io:
Before: [`@ckeditor/ckeditor5-core@40.2.0`](https://arethetypeswrong.github.io/?p=%40ckeditor%2Fckeditor5-core%4040.2.0)
After: [`@ckeditor/ckeditor5-core@0.0.0-nightly-20231218.0`](https://arethetypeswrong.github.io/?p=%40ckeditor%2Fckeditor5-core%400.0.0-nightly-20231218.0) (the warning about CJS is expected, because we only ship ESM)

### publint.dev:
Before: [`@ckeditor/ckeditor5-core@40.2.0`](https://publint.dev/@ckeditor/ckeditor5-core@40.2.0)
After: [`@ckeditor/ckeditor5-core@0.0.0-nightly-20231218.0`](https://publint.dev/@ckeditor/ckeditor5-core@0.0.0-nightly-20231218.0)

**Comment by charlttsie** [comment]
Together with @kubaklatt, we've finished testing the `nightly` packages in a webpack-based setup. We tested both open-source and commercial plugins. Building editors went smoothly and we haven't encountered any issues when building with the `nightly` packages specifically. 

I've also run CKE5 E2E tests using `nightly` plugin versions and no issues have been found there 🎉
